// stylable/dist/src/create-infra-structure.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var cached_process_file_1 = require("./cached-process-file");
var parser_1 = require("./parser");
var stylable_processor_1 = require("./stylable-processor");
var ResolverFactory = require('enhanced-resolve/lib/ResolverFactory');
function createInfrastructure(projectRoot, fileSystem, onProcess, resolveOptions) {
    if (onProcess === void 0) { onProcess = function (x) { return x; }; }
    if (resolveOptions === void 0) { resolveOptions = {}; }
    var eResolver = ResolverFactory.createResolver(__assign({ useSyncFileSystemCalls: true, fileSystem: fileSystem }, resolveOptions));
    var fileProcessor = cached_process_file_1.cachedProcessFile(function (from, content) {
        if (!path.isAbsolute(from)) {
            from = eResolver.resolveSync({}, projectRoot, from);
        }
        return onProcess(stylable_processor_1.process(parser_1.safeParse(content, { from: from })), from);
    }, {
        readFileSync: function (moduleId) {
            if (!path.isAbsolute(moduleId)) {
                moduleId = eResolver.resolveSync({}, projectRoot, moduleId);
            }
            return fileSystem.readFileSync(moduleId, 'utf8');
        },
        statSync: function (moduleId) {
            if (!path.isAbsolute(moduleId)) {
                moduleId = eResolver.resolveSync({}, projectRoot, moduleId);
            }
            var stat = fileSystem.statSync(moduleId);
            if (!stat.mtime) {
                return {
                    mtime: new Date(0)
                };
            }
            return stat;
        }
    });
    return {
        resolvePath: function (context, moduleId) {
            if (!path.isAbsolute(moduleId) && moduleId.charAt(0) !== '.') {
                moduleId = eResolver.resolveSync({}, context, moduleId);
            }
            return moduleId;
        },
        fileProcessor: fileProcessor
    };
}
exports.createInfrastructure = createInfrastructure;
