// stylable/dist/src/stylable.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var bundle_1 = require("./bundle");
var create_infra_structure_1 = require("./create-infra-structure");
var diagnostics_1 = require("./diagnostics");
var parser_1 = require("./parser");
var stylable_processor_1 = require("./stylable-processor");
var stylable_resolver_1 = require("./stylable-resolver");
var stylable_transformer_1 = require("./stylable-transformer");
var Stylable = /** @class */ (function () {
    function Stylable(projectRoot, fileSystem, requireModule, delimiter, onProcess, diagnostics, hooks, scopeRoot, resolveOptions) {
        if (delimiter === void 0) { delimiter = '--'; }
        if (diagnostics === void 0) { diagnostics = new diagnostics_1.Diagnostics(); }
        if (hooks === void 0) { hooks = {}; }
        if (scopeRoot === void 0) { scopeRoot = true; }
        if (resolveOptions === void 0) { resolveOptions = {}; }
        this.projectRoot = projectRoot;
        this.fileSystem = fileSystem;
        this.requireModule = requireModule;
        this.delimiter = delimiter;
        this.onProcess = onProcess;
        this.diagnostics = diagnostics;
        this.hooks = hooks;
        this.scopeRoot = scopeRoot;
        this.resolveOptions = resolveOptions;
        var _a = create_infra_structure_1.createInfrastructure(projectRoot, fileSystem, onProcess, resolveOptions), fileProcessor = _a.fileProcessor, resolvePath = _a.resolvePath;
        this.resolvePath = resolvePath;
        this.fileProcessor = fileProcessor;
        this.resolver = new stylable_resolver_1.StylableResolver(this.fileProcessor, this.requireModule);
    }
    Stylable.prototype.createBundler = function () {
        return new bundle_1.Bundler(this);
    };
    Stylable.prototype.createTransformer = function (options) {
        if (options === void 0) { options = {}; }
        return new stylable_transformer_1.StylableTransformer(__assign({ delimiter: this.delimiter, diagnostics: new diagnostics_1.Diagnostics(), fileProcessor: this.fileProcessor, requireModule: this.requireModule, postProcessor: this.hooks.postProcessor, replaceValueHook: this.hooks.replaceValueHook, scopeRoot: this.scopeRoot }, options));
    };
    Stylable.prototype.transform = function (meta, resourcePath, options) {
        if (options === void 0) { options = {}; }
        if (typeof meta === 'string') {
            var root = parser_1.safeParse(meta, { from: resourcePath });
            meta = new stylable_processor_1.StylableProcessor(new diagnostics_1.Diagnostics()).process(root);
        }
        var transformer = this.createTransformer(options);
        this.fileProcessor.add(meta.source, meta);
        return transformer.transform(meta);
    };
    Stylable.prototype.process = function (fullpath) {
        return this.fileProcessor.process(fullpath);
    };
    return Stylable;
}());
exports.Stylable = Stylable;
