// bit-bin/dist/e2e-helper/e2e-scopes.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

function path() {
  const data = _interopRequireWildcard(require("path"));

  path = function () {
    return data;
  };

  return data;
}

function _os() {
  const data = _interopRequireDefault(require("os"));

  _os = function () {
    return data;
  };

  return data;
}

function _fsExtra() {
  const data = _interopRequireDefault(require("fs-extra"));

  _fsExtra = function () {
    return data;
  };

  return data;
}

function _e2eHelper() {
  const data = require("./e2e-helper");

  _e2eHelper = function () {
    return data;
  };

  return data;
}

// @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
class ScopesData {
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  constructor() {
    (0, _defineProperty2().default)(this, "e2eDir", void 0);
    (0, _defineProperty2().default)(this, "local", void 0);
    (0, _defineProperty2().default)(this, "localPath", void 0);
    (0, _defineProperty2().default)(this, "remote", void 0);
    (0, _defineProperty2().default)(this, "remotePath", void 0);
    (0, _defineProperty2().default)(this, "env", void 0);
    (0, _defineProperty2().default)(this, "envPath", void 0);
    (0, _defineProperty2().default)(this, "globalRemote", void 0);
    (0, _defineProperty2().default)(this, "globalRemotePath", void 0);
    this.e2eDir = path().join(_os().default.tmpdir(), 'bit', 'e2e');
    this.setLocalScope();
    this.setRemoteScope();
    this.setEnvScope();
    this.globalRemote = 'global-remote';
    this.globalRemotePath = path().join(this.e2eDir, this.globalRemote);
  }

  setLocalScope(localScope) {
    this.local = localScope || `${(0, _e2eHelper().generateRandomStr)()}-local`;
    this.localPath = path().join(this.e2eDir, this.local);

    _fsExtra().default.ensureDirSync(this.localPath);
  }

  setRemoteScope() {
    this.remote = `${(0, _e2eHelper().generateRandomStr)()}-remote`;
    this.remotePath = path().join(this.e2eDir, this.remote);
  }

  setEnvScope() {
    this.env = `${(0, _e2eHelper().generateRandomStr)()}-env`;
    this.envPath = path().join(this.e2eDir, this.env);
  }

}

exports.default = ScopesData;
