// bit-bin/dist/consumer/versions-ops/merge-version/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "threeWayMerge", {
  enumerable: true,
  get: function () {
    return _threeWayMerge().default;
  }
});
Object.defineProperty(exports, "mergeVersion", {
  enumerable: true,
  get: function () {
    return _mergeVersion().mergeVersion;
  }
});
Object.defineProperty(exports, "FileStatus", {
  enumerable: true,
  get: function () {
    return _mergeVersion().FileStatus;
  }
});
Object.defineProperty(exports, "getMergeStrategyInteractive", {
  enumerable: true,
  get: function () {
    return _mergeVersion().getMergeStrategyInteractive;
  }
});
Object.defineProperty(exports, "getMergeStrategy", {
  enumerable: true,
  get: function () {
    return _mergeVersion().getMergeStrategy;
  }
});
Object.defineProperty(exports, "MergeOptions", {
  enumerable: true,
  get: function () {
    return _mergeVersion().MergeOptions;
  }
});
Object.defineProperty(exports, "MergeStrategy", {
  enumerable: true,
  get: function () {
    return _mergeVersion().MergeStrategy;
  }
});
Object.defineProperty(exports, "ApplyVersionResults", {
  enumerable: true,
  get: function () {
    return _mergeVersion().ApplyVersionResults;
  }
});
Object.defineProperty(exports, "ApplyVersionResult", {
  enumerable: true,
  get: function () {
    return _mergeVersion().ApplyVersionResult;
  }
});
Object.defineProperty(exports, "FailedComponents", {
  enumerable: true,
  get: function () {
    return _mergeVersion().FailedComponents;
  }
});

function _threeWayMerge() {
  const data = _interopRequireDefault(require("./three-way-merge"));

  _threeWayMerge = function () {
    return data;
  };

  return data;
}

function _mergeVersion() {
  const data = require("./merge-version");

  _mergeVersion = function () {
    return data;
  };

  return data;
}
