// bit-bin/dist/scope/migrations/component-version/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "addBindingPrefixDeclartaion", {
  enumerable: true,
  get: function () {
    return _addBindingPrefix().default;
  }
});
Object.defineProperty(exports, "changeVersionToSemVerDeclartaion", {
  enumerable: true,
  get: function () {
    return _toSemverMigration().default;
  }
});
Object.defineProperty(exports, "changeImportSpecifiersToArray", {
  enumerable: true,
  get: function () {
    return _changeImportSpecifiersToArray().default;
  }
});
Object.defineProperty(exports, "implSpecsToFilesDeclartaion", {
  enumerable: true,
  get: function () {
    return _toFiles().default;
  }
});
Object.defineProperty(exports, "ensureMainFileDeclartaion", {
  enumerable: true,
  get: function () {
    return _ensureMainFile().default;
  }
});
Object.defineProperty(exports, "specsResultstoArrayDeclartaion", {
  enumerable: true,
  get: function () {
    return _specsResultsToArray().default;
  }
});
Object.defineProperty(exports, "updateBindingPrefixToNewDefault", {
  enumerable: true,
  get: function () {
    return _updateBindingPrefixToNewDefault().default;
  }
});
Object.defineProperty(exports, "removeLatestFromCompiler", {
  enumerable: true,
  get: function () {
    return _removeLatestFromCompiler().default;
  }
});

function _addBindingPrefix() {
  const data = _interopRequireDefault(require("./add-binding-prefix"));

  _addBindingPrefix = function () {
    return data;
  };

  return data;
}

function _toSemverMigration() {
  const data = _interopRequireDefault(require("./to-semver-migration"));

  _toSemverMigration = function () {
    return data;
  };

  return data;
}

function _changeImportSpecifiersToArray() {
  const data = _interopRequireDefault(require("./change-import-specifiers-to-array"));

  _changeImportSpecifiersToArray = function () {
    return data;
  };

  return data;
}

function _toFiles() {
  const data = _interopRequireDefault(require("./to-files"));

  _toFiles = function () {
    return data;
  };

  return data;
}

function _ensureMainFile() {
  const data = _interopRequireDefault(require("./ensure-main-file"));

  _ensureMainFile = function () {
    return data;
  };

  return data;
}

function _specsResultsToArray() {
  const data = _interopRequireDefault(require("./specs-results-to-array"));

  _specsResultsToArray = function () {
    return data;
  };

  return data;
}

function _updateBindingPrefixToNewDefault() {
  const data = _interopRequireDefault(require("./update-binding-prefix-to-new-default"));

  _updateBindingPrefixToNewDefault = function () {
    return data;
  };

  return data;
}

function _removeLatestFromCompiler() {
  const data = _interopRequireDefault(require("./remove-latest-from-compiler"));

  _removeLatestFromCompiler = function () {
    return data;
  };

  return data;
}
