// bit-bin/dist/utils/bit/compose-component-path.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.composeComponentPath = composeComponentPath;
exports.composeDependencyPath = composeDependencyPath;
exports.composeDependencyPathForIsolated = composeDependencyPathForIsolated;

function path() {
  const data = _interopRequireWildcard(require("path"));

  path = function () {
    return data;
  };

  return data;
}

function _stringFormat() {
  const data = _interopRequireDefault(require("string-format"));

  _stringFormat = function () {
    return data;
  };

  return data;
}

function _constants() {
  const data = require("../../constants");

  _constants = function () {
    return data;
  };

  return data;
}

function composeComponentPath(bitId, // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
componentsDefaultDirectory = _constants().DEFAULT_COMPONENTS_DIR_PATH) {
  return (0, _stringFormat().default)(componentsDefaultDirectory, {
    name: bitId.name,
    scope: bitId.scope
  });
}

function composeDependencyPath(bitId, // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
dependenciesDir = _constants().DEFAULT_DEPENDENCIES_DIR_PATH) {
  return path().join(dependenciesDir, bitId.toFullPath());
}

function composeDependencyPathForIsolated(bitId, // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
dependenciesDir = _constants().DEFAULT_DEPENDENCIES_DIR_PATH) {
  const getIdPath = () => {
    try {
      return bitId.toFullPath();
    } catch (err) {
      return bitId.name;
    }
  };

  return path().join(dependenciesDir, getIdPath());
}
