// bit-bin/dist/scope/network/exceptions/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ProtocolNotSupported", {
  enumerable: true,
  get: function () {
    return _protocolNotSupported().default;
  }
});
Object.defineProperty(exports, "FsScopeNotLoaded", {
  enumerable: true,
  get: function () {
    return _fsScopeNotLoaded().default;
  }
});
Object.defineProperty(exports, "RemoteScopeNotFound", {
  enumerable: true,
  get: function () {
    return _remoteScopeNotFound().default;
  }
});
Object.defineProperty(exports, "SSHConnectionError", {
  enumerable: true,
  get: function () {
    return _sshConnectionError().default;
  }
});
Object.defineProperty(exports, "PermissionDenied", {
  enumerable: true,
  get: function () {
    return _permissionDenied().default;
  }
});
Object.defineProperty(exports, "NetworkError", {
  enumerable: true,
  get: function () {
    return _networkError().default;
  }
});
Object.defineProperty(exports, "UnexpectedNetworkError", {
  enumerable: true,
  get: function () {
    return _unexpectedNetworkError().default;
  }
});
Object.defineProperty(exports, "SSHInvalidResponse", {
  enumerable: true,
  get: function () {
    return _sshInvalidResponse().default;
  }
});
Object.defineProperty(exports, "AuthenticationFailed", {
  enumerable: true,
  get: function () {
    return _authenticationFailed().default;
  }
});
Object.defineProperty(exports, "OldClientVersion", {
  enumerable: true,
  get: function () {
    return _oldClientVersion().default;
  }
});

function _protocolNotSupported() {
  const data = _interopRequireDefault(require("./protocol-not-supported"));

  _protocolNotSupported = function () {
    return data;
  };

  return data;
}

function _fsScopeNotLoaded() {
  const data = _interopRequireDefault(require("./fs-scope-not-loaded"));

  _fsScopeNotLoaded = function () {
    return data;
  };

  return data;
}

function _remoteScopeNotFound() {
  const data = _interopRequireDefault(require("./remote-scope-not-found"));

  _remoteScopeNotFound = function () {
    return data;
  };

  return data;
}

function _sshConnectionError() {
  const data = _interopRequireDefault(require("./ssh-connection-error"));

  _sshConnectionError = function () {
    return data;
  };

  return data;
}

function _permissionDenied() {
  const data = _interopRequireDefault(require("./permission-denied"));

  _permissionDenied = function () {
    return data;
  };

  return data;
}

function _networkError() {
  const data = _interopRequireDefault(require("./network-error"));

  _networkError = function () {
    return data;
  };

  return data;
}

function _unexpectedNetworkError() {
  const data = _interopRequireDefault(require("./unexpected-network-error"));

  _unexpectedNetworkError = function () {
    return data;
  };

  return data;
}

function _sshInvalidResponse() {
  const data = _interopRequireDefault(require("./ssh-invalid-response"));

  _sshInvalidResponse = function () {
    return data;
  };

  return data;
}

function _authenticationFailed() {
  const data = _interopRequireDefault(require("./authentication-failed"));

  _authenticationFailed = function () {
    return data;
  };

  return data;
}

function _oldClientVersion() {
  const data = _interopRequireDefault(require("./old-client-version"));

  _oldClientVersion = function () {
    return data;
  };

  return data;
}
