// bit-bin/dist/links/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "NodeModuleLinker", {
  enumerable: true,
  get: function () {
    return _nodeModulesLinker().default;
  }
});
Object.defineProperty(exports, "linkAllToNodeModules", {
  enumerable: true,
  get: function () {
    return _linker().linkAllToNodeModules;
  }
});
Object.defineProperty(exports, "getLinksInDistToWrite", {
  enumerable: true,
  get: function () {
    return _linker().getLinksInDistToWrite;
  }
});
Object.defineProperty(exports, "getAllComponentsLinks", {
  enumerable: true,
  get: function () {
    return _linker().getAllComponentsLinks;
  }
});
Object.defineProperty(exports, "reLinkDependents", {
  enumerable: true,
  get: function () {
    return _linker().reLinkDependents;
  }
});
Object.defineProperty(exports, "LinkFile", {
  enumerable: true,
  get: function () {
    return _linkFile().default;
  }
});

function _nodeModulesLinker() {
  const data = _interopRequireDefault(require("./node-modules-linker"));

  _nodeModulesLinker = function () {
    return data;
  };

  return data;
}

function _linker() {
  const data = require("./linker");

  _linker = function () {
    return data;
  };

  return data;
}

function _linkFile() {
  const data = _interopRequireDefault(require("./link-file"));

  _linkFile = function () {
    return data;
  };

  return data;
}
