// bit-bin/dist/consumer/component-ops/add-components/exceptions/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DuplicateIds", {
  enumerable: true,
  get: function () {
    return _duplicateIds().default;
  }
});
Object.defineProperty(exports, "EmptyDirectory", {
  enumerable: true,
  get: function () {
    return _emptyDirectory().default;
  }
});
Object.defineProperty(exports, "IncorrectIdForImportedComponent", {
  enumerable: true,
  get: function () {
    return _incorrectIdImportedComponent().default;
  }
});
Object.defineProperty(exports, "MissingComponentIdForImportedComponent", {
  enumerable: true,
  get: function () {
    return _missingIdImportedComponent().default;
  }
});
Object.defineProperty(exports, "NoFiles", {
  enumerable: true,
  get: function () {
    return _noFiles().default;
  }
});
Object.defineProperty(exports, "PathsNotExist", {
  enumerable: true,
  get: function () {
    return _pathsNotExist().default;
  }
});
Object.defineProperty(exports, "ExcludedMainFile", {
  enumerable: true,
  get: function () {
    return _excludedMainFile().default;
  }
});
Object.defineProperty(exports, "VersionShouldBeRemoved", {
  enumerable: true,
  get: function () {
    return _versionShouldBeRemoved().default;
  }
});
Object.defineProperty(exports, "TestIsDirectory", {
  enumerable: true,
  get: function () {
    return _testIsDirectory().default;
  }
});
Object.defineProperty(exports, "MainFileIsDir", {
  enumerable: true,
  get: function () {
    return _mainFileIsDir().default;
  }
});
Object.defineProperty(exports, "MissingMainFileMultipleComponents", {
  enumerable: true,
  get: function () {
    return _missingMainFileMultipleComponents().default;
  }
});
Object.defineProperty(exports, "PathOutsideConsumer", {
  enumerable: true,
  get: function () {
    return _pathOutsideConsumer().default;
  }
});

function _duplicateIds() {
  const data = _interopRequireDefault(require("./duplicate-ids"));

  _duplicateIds = function () {
    return data;
  };

  return data;
}

function _emptyDirectory() {
  const data = _interopRequireDefault(require("./empty-directory"));

  _emptyDirectory = function () {
    return data;
  };

  return data;
}

function _incorrectIdImportedComponent() {
  const data = _interopRequireDefault(require("./incorrect-id-imported-component"));

  _incorrectIdImportedComponent = function () {
    return data;
  };

  return data;
}

function _missingIdImportedComponent() {
  const data = _interopRequireDefault(require("./missing-id-imported-component"));

  _missingIdImportedComponent = function () {
    return data;
  };

  return data;
}

function _noFiles() {
  const data = _interopRequireDefault(require("./no-files"));

  _noFiles = function () {
    return data;
  };

  return data;
}

function _pathsNotExist() {
  const data = _interopRequireDefault(require("./paths-not-exist"));

  _pathsNotExist = function () {
    return data;
  };

  return data;
}

function _excludedMainFile() {
  const data = _interopRequireDefault(require("./excluded-main-file"));

  _excludedMainFile = function () {
    return data;
  };

  return data;
}

function _versionShouldBeRemoved() {
  const data = _interopRequireDefault(require("./version-should-be-removed"));

  _versionShouldBeRemoved = function () {
    return data;
  };

  return data;
}

function _testIsDirectory() {
  const data = _interopRequireDefault(require("./test-is-directory"));

  _testIsDirectory = function () {
    return data;
  };

  return data;
}

function _mainFileIsDir() {
  const data = _interopRequireDefault(require("./main-file-is-dir"));

  _mainFileIsDir = function () {
    return data;
  };

  return data;
}

function _missingMainFileMultipleComponents() {
  const data = _interopRequireDefault(require("./missing-main-file-multiple-components"));

  _missingMainFileMultipleComponents = function () {
    return data;
  };

  return data;
}

function _pathOutsideConsumer() {
  const data = _interopRequireDefault(require("./path-outside-consumer"));

  _pathOutsideConsumer = function () {
    return data;
  };

  return data;
}
