// bit-bin/dist/scope/exceptions/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ScopeNotFound", {
  enumerable: true,
  get: function () {
    return _scopeNotFound().default;
  }
});
Object.defineProperty(exports, "ScopeJsonNotFound", {
  enumerable: true,
  get: function () {
    return _scopeJsonNotFound().default;
  }
});
Object.defineProperty(exports, "ScopeAlreadyExists", {
  enumerable: true,
  get: function () {
    return _scopeAlreadyExists().default;
  }
});
Object.defineProperty(exports, "SourceNotFound", {
  enumerable: true,
  get: function () {
    return _sourceNotFound().default;
  }
});
Object.defineProperty(exports, "BitNotInScope", {
  enumerable: true,
  get: function () {
    return _bitNotInScope().default;
  }
});
Object.defineProperty(exports, "MergeConflict", {
  enumerable: true,
  get: function () {
    return _mergeConflict().default;
  }
});
Object.defineProperty(exports, "MergeConflictOnRemote", {
  enumerable: true,
  get: function () {
    return _mergeConflictOnRemote().default;
  }
});
Object.defineProperty(exports, "ComponentNotFound", {
  enumerable: true,
  get: function () {
    return _componentNotFound().default;
  }
});
Object.defineProperty(exports, "VersionNotFound", {
  enumerable: true,
  get: function () {
    return _versionNotFound().default;
  }
});
Object.defineProperty(exports, "VersionAlreadyExists", {
  enumerable: true,
  get: function () {
    return _versionAlreadyExists().default;
  }
});
Object.defineProperty(exports, "HashNotFound", {
  enumerable: true,
  get: function () {
    return _hashNotFound().default;
  }
});
Object.defineProperty(exports, "ResolutionException", {
  enumerable: true,
  get: function () {
    return _resolutionException().default;
  }
});
Object.defineProperty(exports, "DependencyNotFound", {
  enumerable: true,
  get: function () {
    return _dependencyNotFound().default;
  }
});
Object.defineProperty(exports, "CorruptedComponent", {
  enumerable: true,
  get: function () {
    return _corruptedComponent().default;
  }
});
Object.defineProperty(exports, "CyclicDependencies", {
  enumerable: true,
  get: function () {
    return _cyclicDependencie().default;
  }
});
Object.defineProperty(exports, "HashMismatch", {
  enumerable: true,
  get: function () {
    return _hashMismatch().default;
  }
});
Object.defineProperty(exports, "InvalidIndexJson", {
  enumerable: true,
  get: function () {
    return _invalidIndexJson().default;
  }
});
Object.defineProperty(exports, "OutdatedIndexJson", {
  enumerable: true,
  get: function () {
    return _outdatedIndexJson().default;
  }
});

function _scopeNotFound() {
  const data = _interopRequireDefault(require("./scope-not-found"));

  _scopeNotFound = function () {
    return data;
  };

  return data;
}

function _scopeJsonNotFound() {
  const data = _interopRequireDefault(require("./scope-json-not-found"));

  _scopeJsonNotFound = function () {
    return data;
  };

  return data;
}

function _scopeAlreadyExists() {
  const data = _interopRequireDefault(require("./scope-already-exists"));

  _scopeAlreadyExists = function () {
    return data;
  };

  return data;
}

function _sourceNotFound() {
  const data = _interopRequireDefault(require("./source-not-found"));

  _sourceNotFound = function () {
    return data;
  };

  return data;
}

function _bitNotInScope() {
  const data = _interopRequireDefault(require("./bit-not-in-scope"));

  _bitNotInScope = function () {
    return data;
  };

  return data;
}

function _mergeConflict() {
  const data = _interopRequireDefault(require("./merge-conflict"));

  _mergeConflict = function () {
    return data;
  };

  return data;
}

function _mergeConflictOnRemote() {
  const data = _interopRequireDefault(require("./merge-conflict-on-remote"));

  _mergeConflictOnRemote = function () {
    return data;
  };

  return data;
}

function _componentNotFound() {
  const data = _interopRequireDefault(require("./component-not-found"));

  _componentNotFound = function () {
    return data;
  };

  return data;
}

function _versionNotFound() {
  const data = _interopRequireDefault(require("./version-not-found"));

  _versionNotFound = function () {
    return data;
  };

  return data;
}

function _versionAlreadyExists() {
  const data = _interopRequireDefault(require("./version-already-exists"));

  _versionAlreadyExists = function () {
    return data;
  };

  return data;
}

function _hashNotFound() {
  const data = _interopRequireDefault(require("./hash-not-found"));

  _hashNotFound = function () {
    return data;
  };

  return data;
}

function _resolutionException() {
  const data = _interopRequireDefault(require("./resolution-exception"));

  _resolutionException = function () {
    return data;
  };

  return data;
}

function _dependencyNotFound() {
  const data = _interopRequireDefault(require("./dependency-not-found"));

  _dependencyNotFound = function () {
    return data;
  };

  return data;
}

function _corruptedComponent() {
  const data = _interopRequireDefault(require("./corrupted-component"));

  _corruptedComponent = function () {
    return data;
  };

  return data;
}

function _cyclicDependencie() {
  const data = _interopRequireDefault(require("./cyclic-dependencie"));

  _cyclicDependencie = function () {
    return data;
  };

  return data;
}

function _hashMismatch() {
  const data = _interopRequireDefault(require("./hash-mismatch"));

  _hashMismatch = function () {
    return data;
  };

  return data;
}

function _invalidIndexJson() {
  const data = _interopRequireDefault(require("./invalid-index-json"));

  _invalidIndexJson = function () {
    return data;
  };

  return data;
}

function _outdatedIndexJson() {
  const data = _interopRequireDefault(require("./outdated-index-json"));

  _outdatedIndexJson = function () {
    return data;
  };

  return data;
}
