// bit-bin/dist/cli/loader/loader-messages.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BEFORE_SHOW_REMOTE = exports.BEFORE_CHECKOUT = exports.BEFORE_STATUS = exports.BEFORE_LOADING_COMPONENTS = exports.BEFORE_EXPORTS = exports.BEFORE_EXPORT = exports.BEFORE_INSTALL_NPM_DEPENDENCIES = exports.BEFORE_PERSISTING_PUT_ON_SCOPE = exports.BEFORE_IMPORT_PUT_ON_SCOPE = exports.BEFORE_RUNNING_SPECS = exports.BEFORE_RUNNING_BUILD = exports.BEFORE_REMOTE_SEARCH = exports.BEFORE_IMPORT_ACTION = exports.BEFORE_REMOTE_UNDEPRECATE = exports.BEFORE_REMOTE_DEPRECATE = exports.BEFORE_REMOVE = exports.BEFORE_MIGRATION = exports.BEFORE_LOCAL_LIST = exports.BEFORE_REMOTE_LIST = exports.BEFORE_IMPORT_ENVIRONMENT = exports.BEFORE_REMOTE_SHOW = void 0;
const BEFORE_REMOTE_SHOW = 'fetching remote component';
exports.BEFORE_REMOTE_SHOW = BEFORE_REMOTE_SHOW;
const BEFORE_IMPORT_ENVIRONMENT = 'importing environment dependencies...';
exports.BEFORE_IMPORT_ENVIRONMENT = BEFORE_IMPORT_ENVIRONMENT;
const BEFORE_REMOTE_LIST = 'listing remote components';
exports.BEFORE_REMOTE_LIST = BEFORE_REMOTE_LIST;
const BEFORE_LOCAL_LIST = 'listing components';
exports.BEFORE_LOCAL_LIST = BEFORE_LOCAL_LIST;
const BEFORE_MIGRATION = 'upgrading working directory...';
exports.BEFORE_MIGRATION = BEFORE_MIGRATION;
const BEFORE_REMOVE = 'removing components';
exports.BEFORE_REMOVE = BEFORE_REMOVE;
const BEFORE_REMOTE_DEPRECATE = 'deprecating remote components';
exports.BEFORE_REMOTE_DEPRECATE = BEFORE_REMOTE_DEPRECATE;
const BEFORE_REMOTE_UNDEPRECATE = 'un-deprecating remote components';
exports.BEFORE_REMOTE_UNDEPRECATE = BEFORE_REMOTE_UNDEPRECATE;
const BEFORE_IMPORT_ACTION = 'importing components';
exports.BEFORE_IMPORT_ACTION = BEFORE_IMPORT_ACTION;

const BEFORE_REMOTE_SEARCH = ({
  scope,
  queryStr
}) => `searching remote scope <${scope}> for '${queryStr}'`;

exports.BEFORE_REMOTE_SEARCH = BEFORE_REMOTE_SEARCH;
const BEFORE_RUNNING_BUILD = 'building components';
exports.BEFORE_RUNNING_BUILD = BEFORE_RUNNING_BUILD;
const BEFORE_RUNNING_SPECS = 'running specs';
exports.BEFORE_RUNNING_SPECS = BEFORE_RUNNING_SPECS;
const BEFORE_IMPORT_PUT_ON_SCOPE = 'importing components';
exports.BEFORE_IMPORT_PUT_ON_SCOPE = BEFORE_IMPORT_PUT_ON_SCOPE;
const BEFORE_PERSISTING_PUT_ON_SCOPE = 'persisting...';
exports.BEFORE_PERSISTING_PUT_ON_SCOPE = BEFORE_PERSISTING_PUT_ON_SCOPE;
const BEFORE_INSTALL_NPM_DEPENDENCIES = 'ensuring package dependencies';
exports.BEFORE_INSTALL_NPM_DEPENDENCIES = BEFORE_INSTALL_NPM_DEPENDENCIES;
const BEFORE_EXPORT = 'exporting component';
exports.BEFORE_EXPORT = BEFORE_EXPORT;
const BEFORE_EXPORTS = 'exporting components';
exports.BEFORE_EXPORTS = BEFORE_EXPORTS;
const BEFORE_LOADING_COMPONENTS = 'loading components';
exports.BEFORE_LOADING_COMPONENTS = BEFORE_LOADING_COMPONENTS;
const BEFORE_STATUS = 'fetching status';
exports.BEFORE_STATUS = BEFORE_STATUS;
const BEFORE_CHECKOUT = 'switching component version...';
exports.BEFORE_CHECKOUT = BEFORE_CHECKOUT;
const BEFORE_SHOW_REMOTE = 'showing a component...';
exports.BEFORE_SHOW_REMOTE = BEFORE_SHOW_REMOTE;
