// bit-bin/dist/api/scope/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "initScope", {
  enumerable: true,
  get: function () {
    return _scopeInit().default;
  }
});
Object.defineProperty(exports, "put", {
  enumerable: true,
  get: function () {
    return _put().default;
  }
});
Object.defineProperty(exports, "fetch", {
  enumerable: true,
  get: function () {
    return _fetch().default;
  }
});
Object.defineProperty(exports, "describeScope", {
  enumerable: true,
  get: function () {
    return _describeScope().default;
  }
});
Object.defineProperty(exports, "catObject", {
  enumerable: true,
  get: function () {
    return _catObject().default;
  }
});
Object.defineProperty(exports, "catComponent", {
  enumerable: true,
  get: function () {
    return _catComponent().default;
  }
});
Object.defineProperty(exports, "scopeList", {
  enumerable: true,
  get: function () {
    return _scopeList().default;
  }
});
Object.defineProperty(exports, "scopeShow", {
  enumerable: true,
  get: function () {
    return _scopeShow().default;
  }
});
Object.defineProperty(exports, "graph", {
  enumerable: true,
  get: function () {
    return _graph().default;
  }
});
Object.defineProperty(exports, "buildInScope", {
  enumerable: true,
  get: function () {
    return _buildInScope().default;
  }
});
Object.defineProperty(exports, "testInScope", {
  enumerable: true,
  get: function () {
    return _testInScope().default;
  }
});
Object.defineProperty(exports, "modifyCIProps", {
  enumerable: true,
  get: function () {
    return _modifyCiProps().default;
  }
});
Object.defineProperty(exports, "ciUpdateAction", {
  enumerable: true,
  get: function () {
    return _ciUpdateAction().default;
  }
});
Object.defineProperty(exports, "scopeConfig", {
  enumerable: true,
  get: function () {
    return _scopeConfig().default;
  }
});
Object.defineProperty(exports, "catScope", {
  enumerable: true,
  get: function () {
    return _catScope().default;
  }
});
Object.defineProperty(exports, "refreshScope", {
  enumerable: true,
  get: function () {
    return _refreshScope().default;
  }
});
Object.defineProperty(exports, "remove", {
  enumerable: true,
  get: function () {
    return _delete().default;
  }
});
Object.defineProperty(exports, "latestVersions", {
  enumerable: true,
  get: function () {
    return _latestVersions().default;
  }
});
Object.defineProperty(exports, "deprecate", {
  enumerable: true,
  get: function () {
    return _deprecation().deprecate;
  }
});
Object.defineProperty(exports, "undeprecate", {
  enumerable: true,
  get: function () {
    return _deprecation().undeprecate;
  }
});

function _scopeInit() {
  const data = _interopRequireDefault(require("./lib/scope-init"));

  _scopeInit = function () {
    return data;
  };

  return data;
}

function _put() {
  const data = _interopRequireDefault(require("./lib/put"));

  _put = function () {
    return data;
  };

  return data;
}

function _fetch() {
  const data = _interopRequireDefault(require("./lib/fetch"));

  _fetch = function () {
    return data;
  };

  return data;
}

function _describeScope() {
  const data = _interopRequireDefault(require("./lib/describe-scope"));

  _describeScope = function () {
    return data;
  };

  return data;
}

function _catObject() {
  const data = _interopRequireDefault(require("./lib/cat-object"));

  _catObject = function () {
    return data;
  };

  return data;
}

function _catComponent() {
  const data = _interopRequireDefault(require("./lib/cat-component"));

  _catComponent = function () {
    return data;
  };

  return data;
}

function _scopeList() {
  const data = _interopRequireDefault(require("./lib/scope-list"));

  _scopeList = function () {
    return data;
  };

  return data;
}

function _scopeShow() {
  const data = _interopRequireDefault(require("./lib/scope-show"));

  _scopeShow = function () {
    return data;
  };

  return data;
}

function _graph() {
  const data = _interopRequireDefault(require("./lib/graph"));

  _graph = function () {
    return data;
  };

  return data;
}

function _buildInScope() {
  const data = _interopRequireDefault(require("./lib/build-in-scope"));

  _buildInScope = function () {
    return data;
  };

  return data;
}

function _testInScope() {
  const data = _interopRequireDefault(require("./lib/test-in-scope"));

  _testInScope = function () {
    return data;
  };

  return data;
}

function _modifyCiProps() {
  const data = _interopRequireDefault(require("./lib/modify-ci-props"));

  _modifyCiProps = function () {
    return data;
  };

  return data;
}

function _ciUpdateAction() {
  const data = _interopRequireDefault(require("./lib/ci-update-action"));

  _ciUpdateAction = function () {
    return data;
  };

  return data;
}

function _scopeConfig() {
  const data = _interopRequireDefault(require("./lib/scope-config"));

  _scopeConfig = function () {
    return data;
  };

  return data;
}

function _catScope() {
  const data = _interopRequireDefault(require("./lib/cat-scope"));

  _catScope = function () {
    return data;
  };

  return data;
}

function _refreshScope() {
  const data = _interopRequireDefault(require("./lib/refresh-scope"));

  _refreshScope = function () {
    return data;
  };

  return data;
}

function _delete() {
  const data = _interopRequireDefault(require("./lib/delete"));

  _delete = function () {
    return data;
  };

  return data;
}

function _latestVersions() {
  const data = _interopRequireDefault(require("./lib/latest-versions"));

  _latestVersions = function () {
    return data;
  };

  return data;
}

function _deprecation() {
  const data = require("./lib/deprecation");

  _deprecation = function () {
    return data;
  };

  return data;
}
