// bit-bin/dist/api/consumer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "init", {
  enumerable: true,
  get: function () {
    return _init().default;
  }
});
Object.defineProperty(exports, "isolate", {
  enumerable: true,
  get: function () {
    return _isolate().default;
  }
});
Object.defineProperty(exports, "remove", {
  enumerable: true,
  get: function () {
    return _remove().default;
  }
});
Object.defineProperty(exports, "deprecate", {
  enumerable: true,
  get: function () {
    return _deprecation().deprecate;
  }
});
Object.defineProperty(exports, "undeprecate", {
  enumerable: true,
  get: function () {
    return _deprecation().undeprecate;
  }
});
Object.defineProperty(exports, "listScope", {
  enumerable: true,
  get: function () {
    return _listScope().listScope;
  }
});
Object.defineProperty(exports, "tagAction", {
  enumerable: true,
  get: function () {
    return _tag().tagAction;
  }
});
Object.defineProperty(exports, "tagAllAction", {
  enumerable: true,
  get: function () {
    return _tag().tagAllAction;
  }
});
Object.defineProperty(exports, "status", {
  enumerable: true,
  get: function () {
    return _status().default;
  }
});
Object.defineProperty(exports, "build", {
  enumerable: true,
  get: function () {
    return _build().build;
  }
});
Object.defineProperty(exports, "buildAll", {
  enumerable: true,
  get: function () {
    return _build().buildAll;
  }
});
Object.defineProperty(exports, "importAction", {
  enumerable: true,
  get: function () {
    return _import().default;
  }
});
Object.defineProperty(exports, "installAction", {
  enumerable: true,
  get: function () {
    return _install().default;
  }
});
Object.defineProperty(exports, "exportAction", {
  enumerable: true,
  get: function () {
    return _export().default;
  }
});
Object.defineProperty(exports, "getConsumerComponent", {
  enumerable: true,
  get: function () {
    return _getConsumerComponent().default;
  }
});
Object.defineProperty(exports, "getScopeComponent", {
  enumerable: true,
  get: function () {
    return _getScopeComponent().default;
  }
});
Object.defineProperty(exports, "test", {
  enumerable: true,
  get: function () {
    return _test().default;
  }
});
Object.defineProperty(exports, "getComponentLogs", {
  enumerable: true,
  get: function () {
    return _getComponentLogs().default;
  }
});
Object.defineProperty(exports, "remoteAdd", {
  enumerable: true,
  get: function () {
    return _remote().add;
  }
});
Object.defineProperty(exports, "remoteList", {
  enumerable: true,
  get: function () {
    return _remote().list;
  }
});
Object.defineProperty(exports, "remoteRm", {
  enumerable: true,
  get: function () {
    return _remote().remove;
  }
});
Object.defineProperty(exports, "config", {
  enumerable: true,
  get: function () {
    return _globalConfig().default;
  }
});
Object.defineProperty(exports, "getDriver", {
  enumerable: true,
  get: function () {
    return _getDriver().default;
  }
});
Object.defineProperty(exports, "watchAll", {
  enumerable: true,
  get: function () {
    return _watch().default;
  }
});
Object.defineProperty(exports, "add", {
  enumerable: true,
  get: function () {
    return _add().addOne;
  }
});
Object.defineProperty(exports, "addMany", {
  enumerable: true,
  get: function () {
    return _add().addMany;
  }
});
Object.defineProperty(exports, "untrack", {
  enumerable: true,
  get: function () {
    return _untrack().default;
  }
});
Object.defineProperty(exports, "unTagAction", {
  enumerable: true,
  get: function () {
    return _untag().default;
  }
});
Object.defineProperty(exports, "move", {
  enumerable: true,
  get: function () {
    return _move().default;
  }
});
Object.defineProperty(exports, "link", {
  enumerable: true,
  get: function () {
    return _link().default;
  }
});
Object.defineProperty(exports, "checkout", {
  enumerable: true,
  get: function () {
    return _checkout().default;
  }
});
Object.defineProperty(exports, "merge", {
  enumerable: true,
  get: function () {
    return _merge().default;
  }
});
Object.defineProperty(exports, "diff", {
  enumerable: true,
  get: function () {
    return _diff().default;
  }
});
Object.defineProperty(exports, "ejectConf", {
  enumerable: true,
  get: function () {
    return _ejectConf().default;
  }
});
Object.defineProperty(exports, "injectConf", {
  enumerable: true,
  get: function () {
    return _injectConf().default;
  }
});
Object.defineProperty(exports, "migrate", {
  enumerable: true,
  get: function () {
    return _migrate().default;
  }
});
Object.defineProperty(exports, "ejectAction", {
  enumerable: true,
  get: function () {
    return _eject().default;
  }
});
Object.defineProperty(exports, "dependencyStatus", {
  enumerable: true,
  get: function () {
    return _dependency_status().default;
  }
});
Object.defineProperty(exports, "login", {
  enumerable: true,
  get: function () {
    return _login().default;
  }
});
Object.defineProperty(exports, "show", {
  enumerable: true,
  get: function () {
    return _show().default;
  }
});
Object.defineProperty(exports, "paintGraph", {
  enumerable: true,
  get: function () {
    return _paintGraph().default;
  }
});

function _init() {
  const data = _interopRequireDefault(require("./lib/init"));

  _init = function () {
    return data;
  };

  return data;
}

function _isolate() {
  const data = _interopRequireDefault(require("./lib/isolate"));

  _isolate = function () {
    return data;
  };

  return data;
}

function _remove() {
  const data = _interopRequireDefault(require("./lib/remove"));

  _remove = function () {
    return data;
  };

  return data;
}

function _deprecation() {
  const data = require("./lib/deprecation");

  _deprecation = function () {
    return data;
  };

  return data;
}

function _listScope() {
  const data = require("./lib/list-scope");

  _listScope = function () {
    return data;
  };

  return data;
}

function _tag() {
  const data = require("./lib/tag");

  _tag = function () {
    return data;
  };

  return data;
}

function _status() {
  const data = _interopRequireDefault(require("./lib/status"));

  _status = function () {
    return data;
  };

  return data;
}

function _build() {
  const data = require("./lib/build");

  _build = function () {
    return data;
  };

  return data;
}

function _import() {
  const data = _interopRequireDefault(require("./lib/import"));

  _import = function () {
    return data;
  };

  return data;
}

function _install() {
  const data = _interopRequireDefault(require("./lib/install"));

  _install = function () {
    return data;
  };

  return data;
}

function _export() {
  const data = _interopRequireDefault(require("./lib/export"));

  _export = function () {
    return data;
  };

  return data;
}

function _getConsumerComponent() {
  const data = _interopRequireDefault(require("./lib/get-consumer-component"));

  _getConsumerComponent = function () {
    return data;
  };

  return data;
}

function _getScopeComponent() {
  const data = _interopRequireDefault(require("./lib/get-scope-component"));

  _getScopeComponent = function () {
    return data;
  };

  return data;
}

function _test() {
  const data = _interopRequireDefault(require("./lib/test"));

  _test = function () {
    return data;
  };

  return data;
}

function _getComponentLogs() {
  const data = _interopRequireDefault(require("./lib/get-component-logs"));

  _getComponentLogs = function () {
    return data;
  };

  return data;
}

function _remote() {
  const data = require("./lib/remote");

  _remote = function () {
    return data;
  };

  return data;
}

function _globalConfig() {
  const data = _interopRequireDefault(require("./lib/global-config"));

  _globalConfig = function () {
    return data;
  };

  return data;
}

function _getDriver() {
  const data = _interopRequireDefault(require("./lib/get-driver"));

  _getDriver = function () {
    return data;
  };

  return data;
}

function _watch() {
  const data = _interopRequireDefault(require("./lib/watch"));

  _watch = function () {
    return data;
  };

  return data;
}

function _add() {
  const data = require("./lib/add");

  _add = function () {
    return data;
  };

  return data;
}

function _untrack() {
  const data = _interopRequireDefault(require("./lib/untrack"));

  _untrack = function () {
    return data;
  };

  return data;
}

function _untag() {
  const data = _interopRequireDefault(require("./lib/untag"));

  _untag = function () {
    return data;
  };

  return data;
}

function _move() {
  const data = _interopRequireDefault(require("./lib/move"));

  _move = function () {
    return data;
  };

  return data;
}

function _link() {
  const data = _interopRequireDefault(require("./lib/link"));

  _link = function () {
    return data;
  };

  return data;
}

function _checkout() {
  const data = _interopRequireDefault(require("./lib/checkout"));

  _checkout = function () {
    return data;
  };

  return data;
}

function _merge() {
  const data = _interopRequireDefault(require("./lib/merge"));

  _merge = function () {
    return data;
  };

  return data;
}

function _diff() {
  const data = _interopRequireDefault(require("./lib//diff"));

  _diff = function () {
    return data;
  };

  return data;
}

function _ejectConf() {
  const data = _interopRequireDefault(require("./lib/eject-conf"));

  _ejectConf = function () {
    return data;
  };

  return data;
}

function _injectConf() {
  const data = _interopRequireDefault(require("./lib/inject-conf"));

  _injectConf = function () {
    return data;
  };

  return data;
}

function _migrate() {
  const data = _interopRequireDefault(require("./lib/migrate"));

  _migrate = function () {
    return data;
  };

  return data;
}

function _eject() {
  const data = _interopRequireDefault(require("./lib/eject"));

  _eject = function () {
    return data;
  };

  return data;
}

function _dependency_status() {
  const data = _interopRequireDefault(require("./lib/dependency_status"));

  _dependency_status = function () {
    return data;
  };

  return data;
}

function _login() {
  const data = _interopRequireDefault(require("./lib/login"));

  _login = function () {
    return data;
  };

  return data;
}

function _show() {
  const data = _interopRequireDefault(require("./lib/show"));

  _show = function () {
    return data;
  };

  return data;
}

function _paintGraph() {
  const data = _interopRequireDefault(require("./lib/paint-graph"));

  _paintGraph = function () {
    return data;
  };

  return data;
}
