// bit-javascript/dist/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.COMPONENT_ORIGINS = exports.SUPPORTED_EXTENSIONS = exports.BIT_MAP = exports.LOCAL_SCOPE_NOTATION = exports.REMOTE_ALIAS_SIGN = exports.INDEX_JS = exports.MODULES_DIR = exports.DEFAULT_BINDINGS_PREFIX = exports.MODULE_NAME = exports.NO_PLUGIN_TYPE = exports.VERSION_DELIMITER = exports.COMPONENTS_DIRNAME = exports.PACKAGE_JSON = exports.LATEST_VERSION = void 0;

/**
 * cache root directory
 */
const LATEST_VERSION = 'latest';
exports.LATEST_VERSION = LATEST_VERSION;
const PACKAGE_JSON = 'package.json';
exports.PACKAGE_JSON = PACKAGE_JSON;
const COMPONENTS_DIRNAME = 'components';
exports.COMPONENTS_DIRNAME = COMPONENTS_DIRNAME;
const VERSION_DELIMITER = '@';
exports.VERSION_DELIMITER = VERSION_DELIMITER;
const NO_PLUGIN_TYPE = 'none';
exports.NO_PLUGIN_TYPE = NO_PLUGIN_TYPE;
const MODULE_NAME = 'bit';
exports.MODULE_NAME = MODULE_NAME;
const DEFAULT_BINDINGS_PREFIX = '@bit';
exports.DEFAULT_BINDINGS_PREFIX = DEFAULT_BINDINGS_PREFIX;
const MODULES_DIR = 'node_modules';
exports.MODULES_DIR = MODULES_DIR;
const INDEX_JS = 'index.js';
exports.INDEX_JS = INDEX_JS;
const REMOTE_ALIAS_SIGN = '@';
exports.REMOTE_ALIAS_SIGN = REMOTE_ALIAS_SIGN;
const LOCAL_SCOPE_NOTATION = '@this';
exports.LOCAL_SCOPE_NOTATION = LOCAL_SCOPE_NOTATION;
const BIT_MAP = '.bit.map.json';
exports.BIT_MAP = BIT_MAP;
const SUPPORTED_EXTENSIONS = ['.js', '.ts', '.jsx', '.tsx', '.css', '.scss', '.less', '.sass', '.vue', '.styl'];
exports.SUPPORTED_EXTENSIONS = SUPPORTED_EXTENSIONS;
const COMPONENT_ORIGINS = {
  IMPORTED: 'IMPORTED',
  AUTHORED: 'AUTHORED',
  NESTED: 'NESTED'
};
exports.COMPONENT_ORIGINS = COMPONENT_ORIGINS;
