// bit-bin/dist/consumer/component/sources/license.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

function _fsExtra() {
  const data = _interopRequireDefault(require("fs-extra"));

  _fsExtra = function () {
    return data;
  };

  return data;
}

function _constants() {
  const data = require("../../../constants");

  _constants = function () {
    return data;
  };

  return data;
}

function _() {
  const data = require(".");

  _ = function () {
    return data;
  };

  return data;
}

class License extends _().AbstractVinyl {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2().default)(this, "override", true);
    (0, _defineProperty2().default)(this, "src", void 0);
  }

  write() {
    // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
    if (!this.override && _fsExtra().default.existsSync(this.path)) return Promise.resolve(); // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
    // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!

    return _fsExtra().default.outputFile(this.path, this.contents);
  }

  serialize() {
    // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
    return this.contents.toString();
  }

  static deserialize(str) {
    // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
    return new License({
      path: _constants().LICENSE_FILENAME,
      contents: str ? Buffer.from(str) : null
    });
  }

}

exports.default = License;
